pysb/simulator/base.py

```diff
@@ -0,0 +1,329 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+import itertools
+import sympy
+from pysb.core import MonomerPattern, ComplexPattern, as_complex_pattern
+
+
+class SimulatorException(Exception):
+    pass
+
+
+class Simulator(object):
+    """An abstract base class for numerical simulation of models.
+
+    Parameters
+    ----------
+    model : pysb.Model
+        Model to simulate.
+    tspan : vector-like, optional
+        Time values over which to simulate. The first and last values define
+        the time range. Returned trajectories are sampled at every value unless
+        the simulation is interrupted for some reason, e.g., due to
+        satisfaction
+        of a logical stopping criterion (see 'tout' below).
+    verbose : bool, optional (default: False)
+        Verbose output.
+
+    Attributes
+    ----------
+    verbose: bool
+        Verbosity flag passed to the constructor.
+    model : pysb.Model
+        Model passed to the constructor.
+    tspan : vector-like
+        Time values passed to the constructor.
+    tout: numpy.ndarray
+        Time points returned by the simulator (may be different from ``tspan``
+        if simulation is interrupted for some reason).
+
+    Notes
+    -----
+    If ``tspan`` is not defined, it may be defined in the call to the
+    ``run`` method.
+
+    The dimensionality of ``tout`` depends on whether a single simulation
+    or multiple simulations are run.
+
+    The dimensionalities of ``y``, ``yobs``, ``yobs_view``, ``yexpr``, and
+    ``yexpr_view`` depend on the number of simulations being run as well
+    as on the type of simulation, i.e., spatial vs. non-spatial.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def __init__(self, model, verbose=False, **kwargs):
+        self.model = model
+        self.verbose = verbose
+        self.tout = None
+        self._y = None
+        self._yobs = None
+        self._yobs_view = None
+        self._yexpr = None
+        self._yexpr_view = None
+        self._yfull = None
+        # Per-run initial conditions/parameter/tspan override
+        self.tspan = kwargs.get('tspan', None)
```

Comment:
Did you avoid making tspan, initials and param_values explicit named args in the method signature so the signature wouldn't be unwieldy for subclasses? I noticed that initials and param_values aren't documented in the class docstring. Does it make sense to allow passing these three things to the constructor, or should we only support setting them on the object after initialization is complete?